### PR TITLE
osd: read some chunks from specified shard for ec pool

### PR DESCRIPTION
**Pull request by XinzeChi**
This idea has been send to ceph-devel.

http://permalink.gmane.org/gmane.comp.file-systems.ceph.devel/23872

Signed-off-by: Xinze Chi <xmdxcxz@gmail.com>

### CODE REVIEW

**Comment by XinzeChi** [comment]
@dachary , we have discuss about this idea some days ago in ceph-devel mail list. 
Could you take a look and review it.
Thanks for any hints.

**Comment by dachary** [comment]
@XinzeChi intsead of adding tests to ensure all cases are handled, another way could be to simplify the implementation and remove border cases. To be more specific, I suggest you split the code in two commits:

* A commit in which ECBackend::objects_read_async removes  (from what is returned by get_min_avail_to_read_shards) the shards that do not need to be read because they contain nothing interesting. And adapt the finish() method to successfully return the desired content when some shards are missing because they are not necessary. It probably is a commit that is not too complicated.
* A commit that further optimizes by only reading the bytes that are necessary within the chunk.

In other words, first optimize to not read chunks that do not contain anything of interest, then optimize to only read the bytes that are of interest within the chunk.

Please forgive me if I missed something, it is entirely possible that I misunderstood part of your patch. I'm quite impressed by the level of understanding you have on this part of the code :-)

**Comment by XinzeChi** [comment]
@dachary , If the only goal is to do your first optimize mentioned above, I think it would be a little simpler. But if we want to cover both the first and second optimize above, I do not think we should split the code in two commits.

**Comment by dachary** [comment]
I think the partial ratio should be separated in a separate pull request. It implements an optimization that is valuable but that makes sense in a context that is larger than just this pull request. As you explained, the ratio can be used to trade bandwidth for speed, if #4312 is merged.

**Comment by dachary** [comment]
@XinzeChi it needs rebasing

**Comment by dachary** [comment]
building under the branch XinzeChi/wip-ec-read on http://ceph.com/gitbuilder.cgi

**Comment by XinzeChi** [comment]
@dachary , gitbuilder just return some compile WARNING.
So could we start a new round of testings for this PR?

**Comment by dachary** [comment]
@XinzeChi I'll run a rados suite on it. Rebuilding on gitbuilders with XinzeChi-wip-ec-read instead of XinzeChi/wip-ec-read because the / seems to cause trouble for teuthology.

**Comment by dachary** [comment]
<pre>
./virtualenv/bin/teuthology-suite --verbose --priority 2000 --suite rados --subset 0/15000 --suite-branch master --machine-type plana,burnupi,mira --distro ubuntu --email loic@dachary.org --ceph XinzeChi-wip-ec-read
</pre>
running at http://pulpito.ceph.com/loic-2015-05-18_23:10:43-rados-XinzeChi-wip-ec-read---basic-multi/

**Comment by dachary** [comment]
** *'mkdir -p -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && cd -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && CEPH_CLI_TEST_DUP_COMMAND=1 CEPH_REF=693af81766abe76676122d278d6575195352770d TESTDIR="/home/ubuntu/cephtest" CEPH_ID="0" PATH=$PATH:/usr/sbin adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 3h /home/ubuntu/cephtest/workunit.client.0/cephtool/test.sh'*
*** "rados/singleton/{all/cephtool.yaml fs/btrfs.yaml msgr-failures/many.yaml}":http://pulpito.ceph.com/loic-2015-05-18_23:10:43-rados-XinzeChi-wip-ec-read---basic-multi/897982
** *wget -O- 'http://git.ceph.com/?p=ceph-qa-chef.git;a=blob_plain;f=solo/solo-from-scratch;hb=HEAD' | CHEF_REPO= CHEF_BRANCH= sh -x*
*** "rados/verify/{1thrash/none.yaml clusters/fixed-2.yaml fs/btrfs.yaml msgr-failures/few.yaml tasks/mon_recovery.yaml validater/valgrind.yaml}":http://pulpito.ceph.com/loic-2015-05-18_23:10:43-rados-XinzeChi-wip-ec-read---basic-multi/898042
*** "rados/verify/{1thrash/default.yaml clusters/fixed-2.yaml fs/btrfs.yaml msgr-failures/few.yaml tasks/rados_cls_all.yaml validater/lockdep.yaml}":http://pulpito.ceph.com/loic-2015-05-18_23:10:43-rados-XinzeChi-wip-ec-read---basic-multi/898018
*** "rados/thrash-erasure-code/{clusters/fixed-2.yaml fs/xfs.yaml msgr-failures/osd-delay.yaml thrashers/morepggrow.yaml workloads/ec-radosbench.yaml}":http://pulpito.ceph.com/loic-2015-05-18_23:10:43-rados-XinzeChi-wip-ec-read---basic-multi/898008
** *'mkdir -p -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && cd -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && CEPH_CLI_TEST_DUP_COMMAND=1 CEPH_REF=693af81766abe76676122d278d6575195352770d TESTDIR="/home/ubuntu/cephtest" CEPH_ID="0" PATH=$PATH:/usr/sbin adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 3h /home/ubuntu/cephtest/workunit.client.0/osdc/stress_objectcacher.sh'*
*** "rados/objectstore/objectcacher-stress.yaml":http://pulpito.ceph.com/loic-2015-05-18_23:10:43-rados-XinzeChi-wip-ec-read---basic-multi/898028

**Comment by dachary** [comment]
Rescheduling the 5 failed tests to get rid of the false negatives
<pre>
run=loic-2015-05-18_23:10:43-rados-XinzeChi-wip-ec-read---basic-multi
eval filter=$(curl --silent http://paddles.front.sepia.ceph.com/runs/$run/ | jq '.jobs[] | select(.status == "dead" or .status == "fail") | .description' | while read description ; do echo -n $description, ; done | sed -e 's/,$//')
./virtualenv/bin/teuthology-suite  --filter="$filter" --priority 2000 --suite rados --subset 0/15000 --suite-branch master --machine-type plana,burnupi,mira --distro ubuntu --email loic@dachary.org --ceph XinzeChi-wip-ec-read
</pre>

**Comment by dachary** [comment]
* http://pulpito.ceph.com/loic-2015-05-20_16:58:45-rados-XinzeChi-wip-ec-read---basic-multi/902093/ is http://tracker.ceph.com/issues/11690
*  http://pulpito.ceph.com/loic-2015-05-20_16:58:45-rados-XinzeChi-wip-ec-read---basic-multi/902096/ is http://tracker.ceph.com/issues/11620

**Comment by XinzeChi** [comment]
@dachary , so we can not continue to test this PR until someone fix this two bug?

**Comment by dachary** [comment]
@xinxinsh these two bugs are not related and do not block this PR. I just spent time on issue 11618 because it was annoying a lot of people and that diverted me from caring for this PR ;-)

**Comment by XinzeChi** [comment]
@dachary , I think you mean me. It does not matter. :-)

**Comment by dachary** [comment]
@XinzeChi I meant you indeed, the autocompletion got me, sorry. @xinxinsh sorry for the noise ;-)

**Comment by dachary** [comment]
@XinzeChi it looks good to me, the rados suite run at 

* http://pulpito.ceph.com/loic-2015-05-18_23:10:43-rados-XinzeChi-wip-ec-read---basic-multi/

with a re-run of environmental failures at 

* http://pulpito.ceph.com/loic-2015-05-20_16:58:45-rados-XinzeChi-wip-ec-read---basic-multi/

show it passes, you added unit tests as well, and you simplified the pull request by removing some superfluous features. I think now is a good time to hand this over to @athanatos who is the ultimate authority to comment and review such a significant change.

**Comment by XinzeChi** [comment]
@dachary , thanks for your reivew and test.
@athanatos , Could you have a look and review?

**Comment by liewegas** [comment]
@XinzeChi what's the status here?

**Comment by XinzeChi** [comment]
@liewegas I am so sorry this PR is delay so long. Coud you review it?

**Comment by XinzeChi** [comment]
@tchaikov, use the small reads which have a size and (offset % stripe_width) distributed randomly on [0, stripe_width) .  this would trigger fast read in ec pool. we need to make sure the chunk overlap case is well handled

**Comment by XinzeChi** [comment]
rebase again

**Comment by liewegas** [comment]
We need to revisit/discuss the codes approach on ceph-devel and reopen this as appropriate.  This shouldn't be so complicated...
